types/json/path_expr.go

```diff
@@ -165,111 +161,217 @@ func (pe PathExpression) ContainsAnyAsterisk() bool {
+	s := &stream{pathExpr: pathExpr, pos: 0}
+	s.skipWhiteSpace()
+	if s.exhausted() || s.read() != '$' {
+		return PathExpression{}, ErrInvalidJSONPath.GenWithStackByArgs(1)
 	}
+	s.skipWhiteSpace()
 
-	pe.legs = make([]pathLeg, 0, len(indices))
+	pe.legs = make([]pathLeg, 0, 16)
 	pe.flags = pathExpressionFlag(0)
 
-	lastEnd := 0
-	for _, indice := range indices {
-		start, end := indice[0], indice[1]
+	var ok bool
+
+	for !s.exhausted() {
+		switch s.peek() {
+		case '.':
+			ok = parseMember(s, &pe)
+		case '[':
+			ok = parseArray(s, &pe)
+		case '*':
+			ok = parseWildcard(s, &pe)
+		default:
+			ok = false
+		}
 
-		// Check all characters between two legs are blank.
-		for i := lastEnd; i < start; i++ {
-			if !isBlank(rune(pathExprSuffix[i])) {
-				err = ErrInvalidJSONPath.GenWithStackByArgs(pathExpr)
-				return
-			}
+		if !ok {
+			return PathExpression{}, ErrInvalidJSONPath.GenWithStackByArgs(s.pos)
 		}
-		lastEnd = end
-
-		if pathExprSuffix[start] == '[' {
-			// The leg is an index of a JSON array.
-			var leg = strings.TrimFunc(pathExprSuffix[start+1:end], isBlank)
-			var indexStr = strings.TrimFunc(leg[0:len(leg)-1], isBlank)
-			var index int
-			if len(indexStr) == 1 && indexStr[0] == '*' {
-				pe.flags |= pathExpressionContainsAsterisk
-				index = arrayIndexAsterisk
-			} else {
-				if index, err = strconv.Atoi(indexStr); err != nil {
-					err = errors.Trace(err)
-					return
-				}
-			}
-			pe.legs = append(pe.legs, pathLeg{typ: pathLegIndex, arrayIndex: index})
-		} else if pathExprSuffix[start] == '.' {
-			// The leg is a key of a JSON object.
-			var key = strings.TrimFunc(pathExprSuffix[start+1:end], isBlank)
-			if len(key) == 1 && key[0] == '*' {
-				pe.flags |= pathExpressionContainsAsterisk
-			} else if key[0] == '"' {
-				// We need unquote the origin string.
-				if key, err = unquoteString(key[1 : len(key)-1]); err != nil {
-					err = ErrInvalidJSONPath.GenWithStackByArgs(pathExpr)
-					return
+
+		s.skipWhiteSpace()
+	}
+
+	if len(pe.legs) > 0 && pe.legs[len(pe.legs)-1].typ == pathLegDoubleAsterisk {
+		return PathExpression{}, ErrInvalidJSONPath.GenWithStackByArgs(s.pos)
+	}
+
+	return
+}
+
+func parseWildcard(s *stream, p *PathExpression) bool {
+	s.skip(1)
+	if s.exhausted() || s.read() != '*' {
+		return false
+	}
+	if s.exhausted() || s.peek() == '*' {
+		return false
+	}
+
+	p.flags |= pathExpressionContainsDoubleAsterisk
+	p.legs = append(p.legs, pathLeg{typ: pathLegDoubleAsterisk})
+	return true
+}
+
+func parseArray(s *stream, p *PathExpression) bool {
+	s.skip(1)
+	s.skipWhiteSpace()
+	if s.exhausted() {
+		return false
+	}
+
+	if s.peek() == '*' {
+		s.skip(1)
+		p.flags |= pathExpressionContainsAsterisk
+		p.legs = append(p.legs, pathLeg{typ: pathLegIndex, arrayIndex: arrayIndexAsterisk})
+	} else {
+		// FIXME: only support an integer index for now. Need to support [last], [1 to 2]... in the future.
+		start := s.pos
+		if !s.incAsLongAs(func(b byte) bool {
```

Comment:
I think it can be restured into `str, ok := s.readWhile(func)`.

Reply:
done, PTAL